core/modules/replica/DatabaseMySQL.cc

```diff
@@ -427,6 +427,67 @@ Connection::Ptr Connection::execute(function<void(Connection::Ptr)> const& scrip
 }
 
 
+Connection::Ptr Connection::executeInOwnTransaction(function<void(Ptr)> const& script,
+                                                    unsigned int maxReconnects,
+                                                    unsigned int timeoutSec,
+                                                    unsigned int maxRetriesOnDeadLock) {
+    string const context =
+        "Connection[" + to_string(_id) + "]::" + string(__func__) + "(_inTransaction=" +
+        to_string(_inTransaction ? 1: 0) +
+        ",maxRetriesOnDeadLock=" + to_string(maxRetriesOnDeadLock) + ")  ";
+
+    auto conn = shared_from_this();
+    try {
+        unsigned int numRetriesOnDeadLock = 0;
+        do {
+            LOGS(_log, LOG_LVL_DEBUG, context << "running user script, numRetriesOnDeadLock: "
+                << numRetriesOnDeadLock);
+            try {
+                return execute(
+                    [&script](Connection::Ptr const& conn) {
+                        conn->begin();
+                        script(conn);
+                        conn->commit();
+                    },
+                    maxReconnects,
+                    timeoutSec
+                );
+            } catch (LockDeadlock const& ex) {
+                if (numRetriesOnDeadLock < maxRetriesOnDeadLock) {
+                    LOGS(_log, LOG_LVL_DEBUG, context << "exception: " << ex.what());
+                    ++numRetriesOnDeadLock;
+                    if (conn->inTransaction()) {
+                        conn->rollback();
+                        conn->begin();
```

Comment:
Because I need to rollback (see `conn-rollback()`) before starting a new transaction. That's exactly what MySQL wanted me to do in the error message (see teh description section of the ticket):
```
Lock wait timeout exceeded; try restarting transaction, errno: 1205
```

Reply:
I understand that you need rollback, my question is why do you need two `begin()` calls?

Reply:
You're right! I don' need the second one. Thank you!